aiida_mlip/calculations/base.py

```diff
@@ -0,0 +1,276 @@
+"""Base class for features common to most calculations."""
+
+from ase.io import read, write
+
+from aiida.common import InputValidationError, datastructures
+import aiida.common.folders
+from aiida.engine import CalcJob, CalcJobProcessSpec
+import aiida.engine.processes
+from aiida.orm import SinglefileData, Str, StructureData
+
+from aiida_mlip.data.config import JanusConfigfile
+from aiida_mlip.data.model import ModelData
+
+
+def validate_inputs(inputs: dict):
+    """
+    Check if the inputs are valid.
+
+    Parameters
+    ----------
+    inputs : dict
+        The inputs dictionary.
+
+    Raises
+    ------
+    ValueError
+        Error message if validation fails, None otherwise.
+    """
+    # Wrapping processes may choose to exclude certain input ports
+    # If the ports have been excluded, skip the validation.
+
+    if not "struct" in inputs and not "config" in inputs:
```

Comment:
it will use struct and not the one in config (which is also what janus does, maybe I should clarify it in the documentation?)

Reply:
I would probably prefer an error raised rather than the situation being silently ignored.

Reply:
I'm not sure I'd consider it being silently ignored, unless I'm misunderstanding. 

Firstly, the config can contain options other than structure, so until we read it later, we won't know at this point whether structure is duplicated or not. If we do want to address this, I think we'd need to do it in the config reader rather than here.

However, in discussions so far we've generally agreed that it can be useful for the config file to be overridable by explicit options, so if the structure is specified in both, using the explicit option is the expected behaviour. As @federicazanca mentioned, this is currently what janus does, but it's open to discussion.

Reply:
You're probably right, I'm just reading this out of context. I guess at the very least it might be worth issuing a log message/warning at the point we do know.